clientconn_test.go

```diff
@@ -689,3 +689,137 @@ func TestDisableServiceConfigOption(t *testing.T) {
 		t.Fatalf("want: method (\"/foo/bar/\") config to be empty, got: %v", m)
 	}
 }
+
+func TestFallbackServiceConfig(t *testing.T) {
+	addr := "passthrough:///non.existent"
+	cc, err := Dial(addr, WithInsecure(), WithFallbackServiceConfig(`{
+    "methodConfig": [
+        {
+            "name": [
+                {
+                    "service": "foo",
+                    "method": "Bar"
+                }
+            ],
+            "waitForReady": true
+        }
+    ]
+}`))
+	if err != nil {
+		t.Fatalf("Dial(%s, _) = _, %v, want _, <nil>", addr, err)
+	}
+	defer cc.Close()
+	time.Sleep(1 * time.Second)
+	m := cc.GetMethodConfig("/foo/Bar")
+	if m.WaitForReady == nil || *m.WaitForReady != true {
+		t.Fatalf("want: method (\"/foo/bar/\") config to be true, got: %v", m)
+	}
+}
+
+func TestFallbackWhenDisableResolverServiceConfigWithoutResolverReturningSC(t *testing.T) {
+	addr := "passthrough:///non.existent"
+	cc, err := Dial(addr, WithInsecure(), WithDisableServiceConfig(), WithFallbackServiceConfig(`{
+    "methodConfig": [
+        {
+            "name": [
+                {
+                    "service": "foo",
+                    "method": "Bar"
+                }
+            ],
+            "waitForReady": true
+        }
+    ]
+}`))
+	if err != nil {
+		t.Fatalf("Dial(%s, _) = _, %v, want _, <nil>", addr, err)
+	}
+	defer cc.Close()
+	time.Sleep(1 * time.Second)
+	m := cc.GetMethodConfig("/foo/Bar")
+	if m.WaitForReady == nil || *m.WaitForReady != true {
+		t.Fatalf("want: method (\"/foo/bar/\") config to be true, got: %v", m)
+	}
+}
+
+func TestFallbackWhenDisableResolverServiceConfigWithResolverReturningSC(t *testing.T) {
+	r, cleanup := manual.GenerateAndRegisterManualResolver()
+	defer cleanup()
+	addr := r.Scheme() + ":///non.existent"
+	cc, err := Dial(addr, WithInsecure(), WithDisableServiceConfig(), WithFallbackServiceConfig(`{
+    "methodConfig": [
+        {
+            "name": [
+                {
+                    "service": "foo",
+                    "method": "Bar"
+                }
+            ],
+            "waitForReady": true
+        }
+    ]
+}`))
+	if err != nil {
+		t.Fatalf("Dial(%s, _) = _, %v, want _, <nil>", addr, err)
+	}
+	defer cc.Close()
+	r.NewServiceConfig(`{
+    "methodConfig": [
+        {
+            "name": [
+                {
+                    "service": "foo",
+                    "method": "Bar"
+                }
+            ],
+            "waitForReady": false
+        }
+    ]
+}`)
+	time.Sleep(1 * time.Second)
```

Comment:
This is hard... how can you confirm that an asynchronous process will never occur?  1s may not be enough, but I don't want to sleep forever, either.

Can we inject the service config synchronously?  This would help all our other tests that have to do polling loops to wait until the service config is installed.

I think this would work:

- `internal/internal.go`: define `var TestingServiceConfigHandled = func() {}`
- `clientconn.go`: `handleServiceConfig` does `defer internal.TestingServiceConfigHandled()`
- Tests reassign `internal.TestingServiceConfig` to something like `func() { close(done) }` and then we know the config is applied after we do `<-done`.  Let's wrap that into a helper function since we have so many tests that update the service config and then poll until it's applied.